pahfit/fitter.py

```diff
@@ -0,0 +1,209 @@
+    the model that they manage for fitting. After this, fit() can be
+    called to apply the model and the astropy fitter to the data. The
+    results will then be retrievable for one component at a time, by
+    passing the component name to get_result().
+
+    """
+
+    @abstractmethod
+    def clear(self):
+        """Reset model.
+
+        After reset, register_() and finalize_model() can be used again.
+
+        """
+        pass
+
+    @abstractmethod
+    def components(self):
+        """Return list of features.
+
+        Only works after finalize_model(). Will return the names passed
+        using the register functions.
+
+        """
+        pass
+
+    @abstractmethod
+    def finalize_model(self):
+        """Process the registered features and prepare for fitting.
+
+        The register functions below allow adding individual features.
+        The exact implementation of how features are added, and
+        finalized in to a single fittable model, depend on the
+        underlying implementation.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_starlight(self, name, temperature, tau):
+        """Register a starlight feature.
+
+        The exact representation depends on the implementation, but the
+        meaning of the parameters should be equivalent.
+
+        Parameters
+        ----------
+        name : str
+            Unique name. Will be used to allow retrieval of the results
+            after the fitting.
+
+        temperature : array of size 3
+            Blackbody temperature, given as [value, lower_bound,
+            upper_bound]. The bounds assume the same system as the
+            features table: if they are masked, the parameter will be
+            fixed, while +inf or -inf mean unbounded.
+
+        tau : array of size 3
+            Analogously, used as power.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_continuum(self, name, temperature, tau):
+        """Register a dust continuum feature."""
+        pass
+
+    @abstractmethod
+    def register_line(self, name, power, wavelength, fwhm):
+        """Register an emission line feature.
+
+        Typically a Gaussian profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_dust_feature(self, name, power, wavelength, fwhm):
+        """Register a dust feature.
+
+        Typically a Drude profile.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_attenuation(self, name, tau):
+        """Register the S07 attenuation component.
+
+        Other types of attenuation might be possible in the future. Is
+        multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_absorption(self, name, tau, wavelength, fwhm):
+        """Register an absorption feature.
+
+        Typically a Drude profile. Is multiplicative.
+
+        """
+        pass
+
+    @abstractmethod
+    def evaluate_model(self, xz):
```

Comment:
Please just `evaluate`.  This is going to make it quite hard to do `tabulate` I guess, since it's just giving "the whole model".  Can you pass whatever wavelengths you want?  How would that work with a model which contains "duplicate" features for JointFitter types?  How does it know which of those features to use?  I strongly suspect this means we need to move this kind of logic out into `tabulate`.

Reply:
Philosophical question: can a `Fitter`, the _same_ `Fitter` you earlier fit data with, be used to `evaluate_model` on a different set of wavelengths?  It has all its own internal ideas of `FWHM` for lines and everything, so in principle it could.  But what about multiple segments, with multiple values for `FWHM` for the same line?  Maybe the rule is: must use same exact bounds (can increase point density if you want), and same shape e.g. (list of N segments).  Enforced, or convention?

Reply:
I do not have a clear answer for this.

For multiple segments, I speculate that there would be multiple model functions, each one to be compared to a certain segment during the fitting (and of which the chi2 are summed together). In that case, and index could be passed, to indicate which part of the "multi-function" should be evaluated.

I think we should first work out how we can reuse the `Fitter` to tabulate individual or groups of components. Perhaps passing a list of feature names to `evaluate()` would be an option? In that case, the evaluation can be generic (implemented in the ABC `Fitter` using the generic component evaluation functions) or specific (making use of the individual astropy components) as necessary.

If we want to avoid duplicating any logic, having it in `Model.tabulate` somehow, or the `Fitter` ABC, would be best.

Reply:
This I think is an argument in favor of implementing the standard set of `Fitter.gaussian/drude/blackbody` etc. components.  There are really only a few of those.  Then we could do away with `Fitter.evaluate` altogether.  Evaluation is not (really) the `Fitter`'s job.  The fitter's job is to do the fit internally using whatever methods and data structures it wants, and return to us the results.  

Beyond that, for `Model`'s convenience, we'd also like each `Fitter` to provide access to the _component functions_ with specified arguments (units, order, etc.) and output units.  Maybe for simplicity and convenience we can ask it for a full `Fitter.attenuation_function`, providing to that method `geometry`, `model`, `tau_si`, and details for other attenuation `tau`'s, by which we can multiply our spectra, or components of our spectra (TBD: increase the `tau` FWHMs for LSF: Sara Profile again).

With this in place, `tabulate` is all we need, and it can be generic.  It just looks at the (set of) spectra it's being called with, does the same parameter marrying with the instrument pack that it does for `Fitter.fit`, and sums up the various `fitter.drude(x,y,z)`, etc., finally multiplying by the `attenuation_function`.

Yes, this is additional work that is broadly similar to what each Fitter is doing internally, but on another level it serves as a good "output comprehension" test.  If you use the `Fitter`'s results to tabulate the model and it doesn't actually match the spectrum, you have not understood the output.

We can even add a test to demand that all `Fitter`'s, when given some standard input, produce the ~same output for each of the component types.